### PR TITLE
Implement Ember.Helper: RFC#53.

### PR DESCRIPTION
**Pull request by mixonic**
Implements https://github.com/emberjs/rfcs/pull/53, new helpers.

For example:

``` js
// app/helpers/full-name.js
import Ember from "ember";

export default Ember.Helper.extend({
  nameBuilder: Ember.inject.service(),
  compute(params) {
    const builder = this.get('nameBuilder');
    return builder.fullName(params[0], params[1]);
  }
});
```

TODO:
- [x] When looking up a helper in dev mode, catch any error about a bare function being registered into the container and rephrase it to be more kind: `A bare function was resolved as a helper: This is not supported. Please upgrade your resolver, or use "Ember.Helper.build"`
- [x] ~~Create an `ember-helper` package?~~
- [x] Can the class based helpers manage their own lifecycle (use the same instance across multiple param and hash updates)?
- [x] export the `Ember.Helper` global
- [x] test usage in subexpressions
- [x] test assertion when used with a block
- [x] revert to `lookupFactory` instead of `lookup` for helpers 
- [x] rebase/squash
- [x] test calling `recompute` on `b` in `(a (b (c)))`
- [x] test teardown of `b` in `(a (b (c)))`

A followup PR should deprecate these legacy APIs in 1.13.1:
- `Ember.Handlebars.makeBoundHelper`
- `Ember.HTMLBars.makeBoundHelper` (private)
- bare functions returned as helpers (`HandlebarsCompatibleHelper`)

### CODE REVIEW

**Comment by mmun** [comment]
Can you add a test that uses a service so we know the container is hooked up correctly?

**Comment by mixonic** [comment]
Tests are green (pending sauce), and I've addressed two items of @mmun feedback:

* Add a test for DI
* Move the helper stream out of `getStream` and into `invokeHelper`

**Comment by mmun** [comment]
Yay :)

File: packages/ember-application/lib/system/application.js [hunk 1]
```diff
@@ -1003,7 +1003,7 @@ Application.reopenClass({
     registry.optionsForType('component', { singleton: false });
     registry.optionsForType('view', { singleton: false });
     registry.optionsForType('template', { instantiate: false });
-    registry.optionsForType('helper', { instantiate: false });
+    registry.optionsForType('helper', { singleton: false });
```
Comment (stefanpenner) [inline packages/ember-application/lib/system/application.js]: I think the singleton part isn't that important. The rest of the system should prefer `lookupFactory(helperName).create(..)` ?

Comment (mixonic) [inline packages/ember-application/lib/system/application.js]: Yep, per the TODO list I'm going to move to using `lookupFactory`.

Comment (stefanpenner) [inline packages/ember-application/lib/system/application.js]: :+1:

**Comment by raytiley** [comment]
Sorry I'm late to the party here. I see this example in the RFC

```js
// app/helpers/can-access.js
import Ember from "ember";

// Usage {{if (can-access 'admin') 'Welcome, boss' 'Heck no!'}}
export default Ember.Helper.extend({
  session: Ember.inject.service(),
  onCurrentUserChange: Ember.observes('session.currentUser', function() {
    this.recompute();
  }),
  compute(params) {
    const currentUser = this.get('session.currentUser');
    return currentUser.can(params[0]);
  }
});
```

I know that under the hood somehow this will use observers but should it be more abstracted so I can avoid owing @stefanpenner 10 bucks each time I write a `Helper`?

Maybe an array property that under the hood will setup an observer and call `recompute` so that the helper would be written like this:

```js
// app/helpers/can-access.js
import Ember from "ember";

// Usage {{if (can-access 'admin') 'Welcome, boss' 'Heck no!'}}
export default Ember.Helper.extend({
  session: Ember.inject.service(),
  dependentKeys: ['session.currentUser'], // When this path changes recompute() is called
  compute(params) {
    const currentUser = this.get('session.currentUser');
    return currentUser.can(params[0]);
  }
});
```

**Comment by mixonic** [comment]
@raytiley an earlier version of the RFC proposed that same API under the property `recomputeObserves`. It was removed with the goal of making helpers as conservative and learnable as possible. It, and other enhancements, can always be added later.

`recompute` is safe to call many times, and will only ever issue a single render. Even if your observer is firing eagerly, and runs several times, the work in `compute` is lazy and will only happen once.

File: packages/ember-htmlbars/tests/compat/handlebars_get_test.js [hunk 2]
```diff
@@ -34,13 +34,13 @@ QUnit.module("ember-htmlbars: compat - Ember.Handlebars.get", {
 QUnit.test('it can lookup a path from the current context', function() {
   expect(1);
 
-  registry.register('helper:handlebars-get', function(path, options) {
+  registry.register('helper:handlebars-get', new HandlebarsCompatibleHelper(function(path, options) {
```
Comment (rwjblue) [inline packages/ember-htmlbars/tests/compat/handlebars_get_test.js]: Seems like this manual wrapping should not be needed.  If it is required does that mean we have broken exporting Handlebars style helpers from `app/helpers/foo-bar.js`?

Comment (rwjblue) [inline packages/ember-htmlbars/tests/compat/handlebars_get_test.js]: After discussing with @mixonic, this is needed because we are wrapping eagerly when a helper is resolved (in the resolver changes in this PR). In this case, we are bypassing the resolver so this wrapping is required.

**Comment by cibernox** [comment]
quick update: Is there any chance of this making it into 1.13?

**Comment by mixonic** [comment]
@cibernox si, it will go into 1.13. Just discussed it again at the f2f. I have one or two small changes but no more blockers, will land it by Sunday eve.

**Comment by cibernox** [comment]
happy

**Comment by rwjblue** [comment]
@cibernox - Yes, it is absolutely critical that we land this into 1.13.

**Comment by mixonic** [comment]
`Ember.Helper` is now feature complete.

I have punted on the ideal life-cycle for class-based helpers. The ideal is that an instance of a class-based helper lasts as long as its value is rendered on the page (even as that value changes), but is destroyed when the value is removed from rendering. Here, the instance is created and destroyed for every computation. This will likely change, but may not change until after 1.13 ships. This fact should not stop anyone from using these APIs.

**Comment by mixonic** [comment]
Will require https://github.com/tildeio/htmlbars/pull/364 and a release for HTMLBars

File: packages/ember-htmlbars/lib/helper.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  isHelper: true,
+  recompute() {
+    this._stream.notify();
+  }
+});
+
+// Ember.Helper.build(function(params, hash) {});
+Helper.reopenClass({
+  isHelperFactory: true,
+  build(helperFn) {
```
Comment (rwjblue) [inline packages/ember-htmlbars/lib/helper.js]: Can you:

* rename this from `build` to `helper`
* move it out of `reopenClass` (because we don't extended class to have a `helper` on their factories)
* export a named `helper` function (for future ease in importing)

File: packages/ember-htmlbars/lib/streams/helper-factory.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  this.linkable = true;
+}
+
+HelperFactoryStream.prototype = create(Stream.prototype);
+
+merge(HelperFactoryStream.prototype, {
+  compute() {
+    let helper = this.helperFactory.create({ _stream: this });
+    let value = helper.compute(getArrayValues(this.params), getHashValues(this.hash));
+    helper.destroy();
```
Comment (wycats) [inline packages/ember-htmlbars/lib/streams/helper-factory.js]: This seems problematic.

File: packages/ember-application/lib/system/resolver.js [hunk 5]
```diff
@@ -374,7 +375,11 @@ export default EmberObject.extend({
     @public
   */
   resolveHelper(parsedName) {
-    return this.resolveOther(parsedName) || helpers[parsedName.fullNameWithoutType];
+    var resolved = this.resolveOther(parsedName) || helpers[parsedName.fullNameWithoutType];
+    if (resolved && !resolved.isHelperFactory && !resolved.isHelperInstance && !resolved.isHTMLBars && typeof resolved === 'function') {
+      resolved = new HandlebarsCompatibleHelper(resolved);
+    }
+    return resolved;
```
Comment (wycats) [inline packages/ember-application/lib/system/resolver.js]: Why does this need to happen ahead of time?

File: packages/ember-views/lib/component_lookup.js [hunk 6]
```diff
@@ -1,14 +1,10 @@
...
+import { CONTAINS_DASH_CACHE } from "ember-htmlbars/system/lookup-helper";
 
 export default EmberObject.extend({
   invalidName(name) {
-    var invalidName = ISNT_HELPER_CACHE.get(name);
-
-    if (invalidName) {
-      Ember.assert(`You cannot use '${name}' as a component name. Component names must contain a hyphen.`);
-    }
+    Ember.assert(`You cannot use '${name}' as a component name. Component names must contain a hyphen.`, CONTAINS_DASH_CACHE.get(name));
```
Comment (wycats) [inline packages/ember-views/lib/component_lookup.js]: This should return true in the assertion case so that unnecessary work is avoided in production mode.

**Comment by cibernox** [comment]
Should I update the blueprint in ember-cli?

I assume that the shorthand function only helper is the preferred approach (for simplicity), but maybe adding a comment with an example how a full-fledged helper is a good idea.

**Comment by mixonic** [comment]
@cibernox I think a PR for that would be fantastic. I'm not sure what the right timing is to merge it into Ember-CLI, but the sooner we can get people off the old APIs the better.
